### ISSUE TITLE
Using Grid with specified capabilities

### ISSUE THREAD

**Issue by maikelgithub** [fhoeben/hsac-fitnesse-fixtures#244]
Hi there,

Not sure if I'm posting this correctly, but here is my current issue:

For our project, I would like to extend our current setup.

This is how we currently start our tests.
```
|script              |selenium driver setup                                                            |
|connect to driver at|${GRID_HUB}|with capabilities|!{browserName:${browser}, version:${FITNESSE_PORT}}|
|show                |driver description                                                               |
```

What we are looking for, is using the information from [the SuiteSetUp from the HSAC project](https://github.com/fhoeben/hsac-fitnesse-fixtures/blob/master/wiki/FitNesseRoot/HsacExamples/SlimTests/BrowserTests/SuiteSetUp.wiki) and combine different instructions into one instruction.

We figured out that working with Chromeoptions directly works fine:
```
|script         |map fixture                            |
|set value      |--start-maximized          |for|args[0]|
|set value      |disable-infobars           |for|args[1]|
|set value      |--disable-extensions       |for|args[2]|
|$chromeProfile=|copy map                               |

|script              |selenium driver setup                       |
|start driver for    |chrome    |with profile   |$chromeProfile   |
|show                |driver description                          |
```

Looking into the code, if we want to use the [Selenium Driver Setup (specifically at line 146)](https://github.com/fhoeben/hsac-fitnesse-fixtures/blob/a64c18d9c4bac8271275c7b089d40be20f0604b5/src/main/java/nl/hsac/fitnesse/fixture/slim/web/SeleniumDriverSetup.java#L146) the driver is started with "Desired Capabilities". Could this be the issue that this is not working the way we want?

Is it possible to make use of Chromium's ChromeOptions, in combination with using the Selenium Grid?

For example:
```
|script         |map fixture                                       |
|set value      |--start-maximized           |for|args[0]          |
|set value      |disable-infobars            |for|args[1]          |
|set value      |--disable-extensions        |for|args[2]          |
|set value      |download.prompt_for_download|to |false|for|args[3]|
|set value      |!{browserName:${browser}}   |for|args[4]          |
|set value      |!{version:${FITNESSE_PORT}} |for|args[5]          |
|$chromeProfile=|copy map                                          |

|script              |selenium driver setup                       |
|connect to driver at|${GRID_HUB}|with capabilities|$chromeProfile|
|show                |driver description                          |
```

To make this possible, we may have a solution for this:
```
public boolean connectToDriverAtWithChromeoptions(String url, Map<String, Object> capabilities)
			throws MalformedURLException {
		DesiredCapabilities desiredCapabilities = new DesiredCapabilities(capabilities);
		// this works from code, preferably this should be set from FitNesse in the SuiteSetUp
		ChromeOptions chromeOptions = new ChromeOptions();
		chromeOptions.addArguments("disable-infobars");
		chromeOptions.addArguments("--start-maximized");
		chromeOptions.addArguments("--disable-extensions");

		Map<String, Object> prefs = new HashMap<String, Object>();
		prefs.put("profile.default_content_settings.popups", 0);
		prefs.put("plugins.always_open_pdf_externally", true);
		prefs.put("download.prompt_for_download", false);
		prefs.put("download.directory_upgrade", true);
		prefs.put("plugins.plugins_disabled","[Chrome PDF Viewer]");

		chromeOptions.setExperimentalOption("prefs", prefs);
		desiredCapabilities.setCapability(ChromeOptions.CAPABILITY, chromeOptions);

		return createAndSetRemoteDriver(url, desiredCapabilities);
	}
```

```
|script                   |selenium driver setup                                                        |
|connect to driver at with|${GRID_HUB}|chromeoptions|!{browserName:${browser}, version:${FITNESSE_PORT}}|
|show                     |driver description                                                           |
```
Any ideas how to make ChromeOptions properly configurable via the Grid?

**Comment by fhoeben** [comment]
I have to admit I haven't looked at this in detail, or tried it myself extensively. But I have tried to start a headless chrome via a grid, and failed.

My current understanding so far is outlined below, but if your code sample works to configure the existing grid node correctly my understanding was obviously incorrect. Do I understand correctly that setting the options from code (as you do in the sample) actually results in a correctly configured node?
(I will have to dig a bit deeper to find whether this is configurable from the wiki, or how we could make it. )

--- Previous idea on how grid capabilities and chrome options were related ---
What I believe is the case is that a grid uses chrome instances/nodes configured a particular way (the node configuration) and with the capabilities you indicate which of these nodes you would like to use. The chromeOptions as you are using them only work for a local chromedriver where they determine how a chrome process is created/launched. So that would mean the chrome options define how chrome works, where in a grid the nodes have a (preconfigured) definition of how they work and capabilities are used to select/find a suitable node. I believe you cannot use chrome options in combination with a grid.

What you would need to do is define a grid node for each of the combinations of options you would like to use in your tests and then for each run select the proper node for that run.
If all your tests use the same options this just means configuring your chrome node correctly, and use that. I have to admit I don't know how to do that, but I expect this is possible.

**Comment by maikelgithub** [comment]
> Do I understand correctly that setting the options from code (as you do in the sample) actually results in a correctly configured node?

We get some random errors. Examples are the test browser starting twice, but every time the browser doesn't stop in the grid. This has to be done manually in the Task Manager or from the Grid. The browser does close, but the process of Chromedriver 32bit.exe doesn't.

> The chromeOptions as you are using them only work for a local chromedriver where they determine how a chrome process is created/launched. So that would mean the chrome options define how chrome works, where in a grid the nodes have a (preconfigured) definition of how they work and capabilities are used to select/find a suitable node.

If we set the chromeOptions in the Grid, we get the following setup in the node config and SuiteSetUp in FitNesse:
```
{
      "browserName": "chrome",
      "maxInstances": 5,
      "version": "8900",      
      "goog:chromeOptions": {
        "args": ["--disable-extensions", "disable-infobars", "--start-maximized"]
      }
    },

|script              |selenium driver setup                                                            |
|connect to driver at|${GRID_HUB}|with capabilities|!{browserName:${browser}, version:${FITNESSE_PORT}}|
|show                |driver description                                                               |
```
After connecting to the Grid, the chromeOptions are visible:
```
capabilities: Capabilities {browserName: chrome, goog:chromeOptions: {args: [--disable-extensions, disable-infobars, --start-maximized]}, maxInstances: 5, platform: XP, platformName: XP, seleniumProtocol: WebDriver, server:CONFIG_UUID: ab7d6999-71ff-4bc2-b98c-3d0..., version: 8900}
```

--------------------------------------------------------------------------------
At this point we start the test. Unfortunately, looking at the webdrivers window, we see that FitNesse overwrites the capabilities set and continues with that part. 
```
INFO [ActiveSessionFactory.apply] - Capabilities are: {
  "browserName": "chrome",
  "version": "8900"
}
```

**Comment by fhoeben** [comment]
And have you tried just adding the `"goog:chromeOptions"` key from the wiki (similar to what you do in the node config)?
So the value would be a nested map with single key `args` which has a list with three values?

So something like:
```
|script   |map fixture                                        |
|set value|chrome              |for|browserName               |
|set value|${FITNESSE_PORT}    |for|version                   |
|set value|--disable-extensions|for|goog:chromeOptions.args[0]|
|set value|disable-infobars    |for|goog:chromeOptions.args[1]|
|set value|--start-maximized   |for|goog:chromeOptions.args[2]|
|$capa=   |copy map                                           |

|script              |selenium driver setup              |
|connect to driver at|${GRID_HUB}|with capabilities|$capa|
|show                |driver description                 |
```

**Comment by maikelgithub** [comment]
Thank you. The chromeOptions work according to your description. 

For the second part, would it work similarly?
```
Map<String, Object> prefs = new HashMap<String, Object>();
		prefs.put("profile.default_content_settings.popups", 0);
		prefs.put("plugins.always_open_pdf_externally", true);
		prefs.put("download.prompt_for_download", false);
		prefs.put("download.directory_upgrade", true);
		prefs.put("plugins.plugins_disabled","[Chrome PDF Viewer]");
		prefs.put("download.default_directory", "C:\\SeleniumTests\\PDF");
		prefs.put("profile.default_content_settings.popups", 0);

		chromeOptions.setExperimentalOption("prefs", prefs);
		desiredCapabilities.setCapability(ChromeOptions.CAPABILITY, chromeOptions);
```

We've tried similar things according to your earlier advice. Unfortunately with no succes so far.

**Comment by maikelgithub** [comment]
Looking further into this, we haven't found any solution yet for setting the download pdf option to on. It's possible in other languages. Maybe it's not possible in FitNesse.

For anyone who is looking for a solution, we wrote the following script in FitNesse. Using the website of Van Lanschot Bankiers as demo:

```
|script                                                                                                                                         |
|open          |https://www.vanlanschotkempen.com/nl/financieel/financiele-resultaten/2018                                                      |
|execute script|var q = document.querySelectorAll('a'); for (var i = 0; i < q.length; i++) { q[i].setAttribute("download","fitnesseDownload") };|
|click         |!-Persbericht Q3 trading update - 2 november 2018-!                                                                             |
|wait          |3                                                                   |seconds                                                    |
|show          |take screenshot                                                     |testje1                                                    |
|refresh                                                                                                                                        |
|click         |!-Persbericht Q3 trading update - 2 november 2018-!                                                                             |
|wait          |3                                                                   |seconds                                                    |
|show          |take screenshot                                                     |testje2                                                    |
```

The first time, FitNesse will download the link as a pdf. The second time, the PDF document will be opened in a new tab inside the Chrome browser started by FitNesse.

Thank you for all your help so far @fhoeben

**Comment by fhoeben** [comment]
Is there a particular reason you can't download the PDF using the 'download' command of BrowserTest 

Something like:

```
|script|browser test                                                              |
|open  |https://www.vanlanschotkempen.com/nl/financieel/financiele-resultaten/2018|
|show  |download         |Persbericht Q3 trading update - 2 november 2018         |

```

When I use this I get the PDF as file inside the fitnesse test report, and you could possibly manipulate it further using 'file fixture' or '[pdf fixture](https://github.com/fhoeben/hsac-fitnesse-pdf/blob/master/src/main/java/nl/hsac/fitnesse/fixture/slim/PdfFixture.java)' of  https://github.com/fhoeben/hsac-fitnesse-pdf. 

To manipulate it further you should store download's result in a variable (instead of doing 'show'):
```
|script|browser test                                                              |
|open  |https://www.vanlanschotkempen.com/nl/financieel/financiele-resultaten/2018|
|$file=|download         |Persbericht Q3 trading update - 2 november 2018         |

|script|file fixture       |
|check |size of|$file|>2048|
```
<img width="962" alt="screenshot 2018-11-29 at 15 38 45" src="https://user-images.githubusercontent.com/5467859/49229011-eaaf7980-f3ec-11e8-9e8a-ea2cd774d3f1.png">

**Comment by maikelgithub** [comment]
Unfortunately, this function does not work. The errors are:

> java.lang.RuntimeException: Unable to get response from: https://www.vanlanschotkempen.com/media/3148/persbericht-trading-update-q3-van-lanschot-kempen.pdf

And: 
> Caused by: org.apache.http.conn.HttpHostConnectException: Connect to www.vanlanschotkempen.com:443 [www.vanlanschotkempen.com/185.151.220.40] failed: Connection timed out: connect

I think that our policies block us from using Java to download files. This is why we were looking for alternative methods.

**Comment by fhoeben** [comment]
Sounds like the machine running the test/fitnesse is not allowed to download from the site and your selenium node is. I posted the screenshot from my machine so the problem is not with the fixtures or the target site, it must be the environment you're running your tests from.

Is this the actual site you are testing? If so, why is the Java process not allowed to connect to it? Or is this a silly question?

**Comment by maikelgithub** [comment]
This website is a demonstration. We test our applications running inside our network.

The Java processes are all running on the same machine. The Chrome process is allowed to connect to the internet, but Java is not. This is due to Chrome automatically taking the system proxy settings. As Java is running locally to connect to the Grid and further towards the ChromeDriver, that part is fine. The Chrome Browser that opened by FitNesse has the correct proxy settings so all downloads from there are fine.

I just completed a different approach. Setting AlwaysOpenPdfExternally as registry for all users on our machines. I would have preferred the method of FitNesse as we now miss a lot of extra functionality from the PdfFixture. I'm sure we can write some code to make it work for our setup as well.

**Comment by fhoeben** [comment]
I believe you can configure Java to use your system's proxy also, by specifying some system properties when launching Java. Although I have to admit I haven't done that myself.
Such system properties, when present, should be respected by the fitnesse code downloading.

But if the site you used was only for demonstration purposes: is the fixture code also not allowed to connect to the actual site you want to use in your test? Since that is inside your own network I assume no proxy is needed, so the connection should be allowed...

**Comment by fhoeben** [comment]
I believe the proxy used for downloading files (which is actually done via Apache http client) can be set using the following system properties:
 *  http.proxyHost
 *  http.proxyPort
 *  https.proxyHost
 *  https.proxyPort
 *  http.nonProxyHosts

**Comment by maikelgithub** [comment]
Dear @fhoeben 

thank you for all your help so far.

Looking into this issue , I discovered that setting the settings you described don't work for our testenvironment.

However, looking into [this](https://github.com/fhoeben/hsac-fitnesse-fixtures/issues/128) explanation we managed to fix the issue with our networkenvironment. Now the download function of FitNesse works again like it should.

Once again thank you for all your help so far. This issue may now be closed.
